File: tree/validate_binary_search.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) 
    {
        
        return isValidBST(root, null, null);
    }
        

    boolean isValidBST(TreeNode root, TreeNode left, TreeNode right)
    {
        // helper function;
        if(root == null)
        {
            return true;
        }
        if(left!=null && left.val >= root.val)
        {
            return false;
        }
        if(right!=null && right.val <= root.val)
        {
            return false;
        }
        return isValidBST(root.left, left, root) && isValidBST(root.right, root, right);
    }   
}

File: easy/tree/validate_binary_search.cpp
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, NULL, NULL);
    }

    bool isValidBST(TreeNode* root, TreeNode* min_node, TreeNode* max_node) {
        if (root == nullptr) return true;
        if (min_node && root->val <= min_node->val || max_node && root->val >= max_node->val)
            return false;
        return isValidBST(root->left, min_node, root) && isValidBST(root->right, root, max_node);
    }
};